### DISCUSSION TITLE
Custom dialogs

### DISCUSSION THREAD

**Discussion by stevebeisner** [beeware/toga#2763] [Q&A] [answered]
I'm porting an app using another language/gui framework, that is structured as a series of custom dialogs, each dialog followed by a gui-independent action.   Each custom dialog consists of a variety of widgets. Each dialog sets or returns values when it is completed and the dialog is closed.   I suppose creating such custom dialogs in Toga is possible, perhaps based on Window (?), but I'm not sure what such an implementation would look like.  Any suggestions, pointers, or examples? Thanks!

**Comment by freakboy3742** [comment, marked as answer]
Your instinct is correct - the best approach at present would be an instance of Window.

Create the instance of Window, possibly with `closable=False` and/or `resizable=False` (to ensure that the user can't dismiss the window, or make it a weird size). Add window content matching your widget's input requirement, adding an "OK" button (or whatever acceptance mechanism you want) to the bottom of the window. Add a handler to the OK button that does whatever acceptance handling is required, plus possibly invokes close on the window. Show the window with `window.show()`.

If you want, you can wrap up your custom dialog as a subclass of Window; for extra bonus points, you can make the subclass `await`able so that you can use your dialog in a similar way to other dialogs in Toga. A sample implementation that has paid no attention to actual widget layout, just puts a text input and a button in a dialog:
```
class SampleDialog(toga.Window):
    def __init__(self):
        super().__init__(title="My Dialog", resizable=False, size=(400, 300))

        self.textinput = toga.TextInput()
        self.ok_button = toga.Button("OK", on_press=self.on_accept)
        self.content = toga.Box(children=[self.textinput, self.ok_button])
        self.future = self.app.loop.create_future()

    def on_accept(self, widget, **kwargs):
        self.future.set_result(self.textinput.value)
        self.close()

    def __await__(self):
        return self.future.__await__()
```

Displaying the dialog can then be triggered with:
```
        dialog = SampleDialog()
        dialog.show()
        result = await dialog
        print("Dialog result is {result}")
```
This will show the dialog, then wait for the OK button to be pressed; when it is, the print statement will execute.

**Reply by stevebeisner** [reply]
Perfect!

**Reply by ndaskalov-mms** [reply]
Hi, I'm new to toga and not very experienced in Python as well. I created simple function to test the approach above, but got the following error. Can you please  suggest what I'm doing wrong.
```
Error description:
  File "c:\source\test1.py", line 31
    result = await dialog
             ^^^^^^^^^^^^
SyntaxError: 'await' outside async function
```
Here is the code:
```python
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW

class HelloWorld(toga.App):

    class SampleDialog(toga.Window):
        def __init__(self):
            super().__init__(title="My Dialog", resizable=False, size=(400, 300))
            self.textinput = toga.TextInput()
            self.ok_button = toga.Button("OK", on_press=self.on_accept)
            self.content = toga.Box(children=[self.textinput, self.ok_button])
            self.future = self.app.loop.create_future()

    def on_accept(self, widget, **kwargs):
        self.future.set_result(self.textinput.value)
        self.close()

    def __await__(self):
        return self.future.__await__()

    def startup(self):   
        main_box = toga.Box(style=Pack(direction=COLUMN))
        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()
        
        dialog = self.SampleDialog()
        dialog.show()
        result = await dialog
        print("Dialog result is {result}")

def main():
    return HelloWorld('Dialog test', 'org.example.twowintst')

if __name__ == "__main__":
    main().main_loop()
```

**Reply by freakboy3742** [reply]
@ndaskalov-mms You've missed a critical detail in the process of embedding the class inside another class - the `on_accept` and `__await__` methods have been defined as methods of the *app*, not methods of the dialog. If you indent those two methods (or, better still - don't make the SampleDialog class an inner class), you should be able to await the dialog instance.

**Reply by ndaskalov-mms** [reply]
Thanks for the explanation, my knowledge is really scarce. I refactored on_accept and __await__ as methods of the SampleDialog class and put it outside of outside of the HelloWorld class.
But the problem is still the same - await can be used only in async def function. In the case, shall I made startup function async, or main?
Please pardon for my ignorance😒

**Reply by freakboy3742** [reply]
My apologies - I missed the detail that the dialog was being displayed in the startup method.

The startup method cannot be made async - it needs to be run in a synchronously because of where it occurs in the app's lifecycle. If you want to display the dialog as soon as the app starts, you need to add an `on_running` handler - that is, a second method, in addition to startup, that is defined as `async def on_running(self):` that is only used to show the dialog on startup. This method will be invoked exactly once, as soon as the app's event loop is running.

**Reply by ndaskalov-mms** [reply]
10x a lot!

**Comment by hemna** [comment]
unfortunately this doesn't work on ios as it doesn't support secondary windows.

**Reply by freakboy3742** [reply]
Sure - that's why I said "the best approach *at present*". There's absolutely a need to be able to customise dialogs; right now, the only viable workaround (other than platform-native code) is a custom window.
